robots/cmd/flake-report-creator/cmd/jenkins.go

```diff
@@ -0,0 +1,444 @@
+	}
+	jLog = log.StandardLogger().WithField("type", "jenkins")
+)
+
+func init() {
+	opts = jenkinsOptions{}
+	jenkinsCommand.PersistentFlags().StringVar(&opts.endpoint, "endpoint", defaultJenkinsBaseUrl, "jenkins base url")
+	jenkinsCommand.PersistentFlags().StringVar(&opts.jobNamePattern, "jobNamePattern", defaultJenkinsJobNamePattern, "jenkins job name pattern to filter jobs for for the report (note the pattern that is used for injecting the CNV version)")
+	jenkinsCommand.PersistentFlags().StringVar(&opts.cnvVersions, "cnvVersions", defaultCNVVersions, "comma separated list of cnv versions to report")
+	jenkinsCommand.PersistentFlags().DurationVar(&opts.startFrom, "startFrom", 14*24*time.Hour, "The duration when the report data should be fetched")
+	jenkinsCommand.PersistentFlags().IntVar(&opts.maxConnsPerHost, "maxConnsPerHost", 5, "The maximum number of connections to the endpoint (to avoid getting rate limited)")
+	fileNameRegex = regexp.MustCompile("^(partial\\.)junit\\.functest(\\.1)\\.xml$")
+}
+
+type jenkinsOptions struct {
+	endpoint        string
+	jobNamePattern  string
+	startFrom       time.Duration
+	cnvVersions     string
+	maxConnsPerHost int
+}
+
+type JenkinsReportParams struct {
+	flakefinder.Params
+	JenkinsBaseURL string
+}
+
+func JenkinsCommand() *cobra.Command {
+	return jenkinsCommand
+}
+
+func runJenkinsReport(cmd *cobra.Command, args []string) error {
+	err := cmd.InheritedFlags().Parse(args)
+	if err != nil {
+		return err
+	}
+
+	if err = globalOpts.Validate(); err != nil {
+		_, err2 := fmt.Fprint(cmd.OutOrStderr(), cmd.UsageString(), err)
+		if err2 != nil {
+			return err2
+		}
+		return fmt.Errorf("invalid arguments provided: %v", err)
+	}
+
+	cnvVersions := strings.Split(opts.cnvVersions, ",")
+
+	ctx := context.Background()
+
+	// limit http client connections to avoid 504 errors, looks like we are getting rate limited
+	client := &http.Client{
+		Transport: &http.Transport{
+			MaxConnsPerHost: opts.maxConnsPerHost,
+		},
+	}
+
+	jLog.Printf("Creating client for %s", opts.endpoint)
+	jenkins := gojenkins.CreateJenkins(client, opts.endpoint)
+	_, err = jenkins.Init(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to contact jenkins %s: %v", opts.endpoint, err)
+	}
+
+	jLog.Printf("Fetching jobs")
+	jobNames, err := jenkins.GetAllJobNames(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch jobs: %v", err)
+	}
+	jLog.Printf("Fetched %d jobs", len(jobNames))
+
+	startOfReport := time.Now().Add(-1 * opts.startFrom)
+	endOfReport := time.Now()
+
+	junitReportsFromMatchingJobs := fetchJunitReportsFromMatchingJobs(startOfReport, endOfReport, cnvVersions, jobNames, jenkins, ctx)
+	writeReportToFile(startOfReport, endOfReport, junitReportsFromMatchingJobs)
+
+	return nil
+}
+
+func fetchJunitReportsFromMatchingJobs(startOfReport time.Time, endOfReport time.Time, cnvVersions []string, jobNames []gojenkins.InnerJob, jenkins *gojenkins.Jenkins, ctx context.Context) []*flakefinder.JobResult {
```

Comment:
Consider the following slightly changed implementation to avoid a few pitfalls:

```go
func fetchJunitReportsFromMatchingJobs(startOfReport time.Time, endOfReport time.Time, cnvVersions []string, jobNames []gojenkins.InnerJob, jenkins *gojenkins.Jenkins, ctx context.Context) []*flakefinder.JobResult {
	reports := []*flakefinder.JobResult{}

	reportChan := make(chan []*flakefinder.JobResult)
	var wg sync.WaitGroup

        wg.Add(len(cnvVersions) * len(jenkinsJobs))
	for _, cnvVersion := range cnvVersions {
		jLog.Printf("Filtering jobs for CNV %s matching %s", cnvVersion, opts.jobNamePattern)
		compile, err := regexp.Compile(fmt.Sprintf(opts.jobNamePattern, cnvVersion))
		if err != nil {
			jLog.Fatalf("failed to fetch jobs: %v", err)
		}
		for _, jobName := range jobNames {
			if !compile.MatchString(jobName.Name) {
				continue
			}

			go func(innerJob gojenkins.InnerJob, ctx context.Context, fLog *log.Entry) {
                               defer wg.Done()
				fLog.Printf("Fetching job")
				job, err := jenkins.GetJob(ctx, innerJob.Name)
				if err != nil {
					fLog.Fatalf("failed to fetch job: %v", err)
				}

				lastBuild, err := job.GetLastBuild(ctx)
				if err != nil {
					fLog.Fatalf("failed to fetch last build: %v", err)
				}

				completedBuilds := fetchCompletedBuildsForJob(startOfReport, endOfReport, lastBuild, job, ctx, fLog)
				junitFilesFromArtifacts := fetchJunitFilesFromArtifacts(completedBuilds, fLog)
				reportsPerJob := convertJunitFileDataToReport(junitFilesFromArtifacts, ctx, job, fLog)

				reportChan <- reportsPerJob
			}(jobName, ctx, jLog.WithField("job", jobName.Name))
		}
	}
       go func() {
          defer close(reportChan)
           wg.Wait()
       }

	for reportsPerJob := range reportChan {
		reports = append(reports, reportsPerJob...)
	}
	return reports
}
```

Reply:
Oh I just see that the jobs are not filtered yet. That would have to be done before the wg.Add(), otherwise we may block indefinitely.

Reply:
@rmohr thank you for you suggestions, greatly appreciate that. I have adjusted the code as you suggested, PTAL!